Aj/tnl 6743 group config enrollment partition

staubina opened the pull request:
## [TNL-6743](https://openedx.atlassian.net/browse/TNL-6743)

### Description

This PR adds the Enrollment Track content groups to the Studio Content Group configuration page when the feature is enabled. 

All Enrollment track groups need to be added in Course Modes and also they should not be editable in the Configuration section.

### Sandbox
- [x] Build a sandbox for your branch and add a link here
https://staubina-enrollment.sandbox.edx.org
https://studio-staubina-enrollment.sandbox.edx.org

2 Courses are available for testing
AuditOnly
AuditAndVerified

### Testing
- [x] i18n
- [x] RTL
- [x] safecommit violation code review process
- [x] Unit, integration, acceptance tests as appropriate

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] Code review: @cahrens 
- [x] Code review: @jlajoie 
- [x] Doc Review: @catong 

FYI: @sstack22 

### Post-review
- [ ] Rebase and squash commits

staubina commented:
jenkins run js

cahrens commented on cms/templates/group_configurations.html:
```diff
@@ -47,30 +49,40 @@ <h1 class="page-header">
       <article class="content-primary" role="main">
+        % if should_show_enrollment_track:
+          <div class="wrapper-groups enrollment-tracks">
+            <h3 class="title">${_("Enrollment Tracks")}</h3>
```
Get the name from the UserPartition (both here below-- "Content Groups").

cahrens commented on cms/templates/group_configurations.html:
```diff
@@ -47,30 +49,40 @@ <h1 class="page-header">
     <section class="content">
+
       <article class="content-primary" role="main">
+        % if should_show_enrollment_track:
```
I'd love to see this template just iterate over user partitions (excluding A/B experiments, which are special cased) instead of having if blocks for enrollment tracks and content groups.

staubina commented:
jenkins run bokchoy

staubina commented:
jenkins run javascript

staubina commented:
jenkins run js

sstack22 commented:
👍  from me! :) Out of scope, but why do we should the group number?

cahrens commented:
Need to change the "content group" language here:

![image](https://cloud.githubusercontent.com/assets/484484/24971036/7aa95a22-1f84-11e7-93ce-21f8c7486ee9.png)

cahrens commented:
I don't think of Enrollment Modes as Content Groups. Content groups can be linked to Cohorts, Enrollment Modes cannot. @catong, help?

![image](https://cloud.githubusercontent.com/assets/484484/24971284/6d95204a-1f85-11e7-8796-00956a619c98.png)

catong commented:
@cahrens Yep, @staubina volunteered to do a first draft of the text; I need to find time to review and make suggestions.

cahrens commented on cms/djangoapps/contentstore/course_group_config.py:
```diff
@@ -286,18 +285,19 @@ def _iterate_items_and_content_group_ids(course, items):
         Iterate through items and content group IDs in a course.
 
-        This will yield group IDs *only* for cohort user partitions.
+        This will yield group IDs for all user partitions.
```
Except the random partition.

cahrens commented on cms/djangoapps/contentstore/course_group_config.py:
```diff
@@ -363,3 +363,21 @@ def get_or_create_content_group(store, course):
         return content_group_configuration
+
+    @staticmethod
+    def get_all_content_groups(store, course):
```
I would make this get_all_partition_groups, or get_all_user_partition_groups, or even get_all_group_configuration_groups.

We need to settle whether or not an enrollment track group is a "content group". Enrollment Tracks are a type of UserPartition with scheme EnrollmentTrackPartitionScheme. Content Groups are a type of UserPartition with scheme CohortPartitionScheme. UserPartitions have Groups.

cahrens commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -1521,15 +1521,32 @@ def group_configurations_list_handler(request, course_key_string):
+                        all_configurations.insert(0, config)
+
+            # Add empty content group if list returned is empty.  This will add ability to add new groups in the view.
+            if len(all_configurations) == 0 or COHORT_SCHEME not in group_schemes:
```
This can be simplified to just COHORT_SCHEME not in group_schemes. If there is no Content Group Configuration, we want to create an empty one.

cahrens commented on docs/cms_config.ini:
```diff
@@ -39,6 +39,7 @@ login = getting_started/index.html
 register = getting_started/index.html
 content_libraries = course_components/libraries.html
 content_groups = course_features/cohorts/cohorted_courseware.html
+enrollment_tracks = course_features/cohorts/cohorted_courseware.html
```
I don't think this should point here.

staubina replied:
I meant to add a comment here.  This was the location @sstack22 pointed me to initially. She said there is work in Docs to update this section. We planned to change this in the future. 

FYI @sstack22 @catong

cahrens commented on cms/static/js/views/content_group_details.js:
```diff
@@ -41,7 +42,8 @@ define([
                 showContentGroupUsages: showContentGroupUsages || false,
-                HtmlUtils: HtmlUtils
+                HtmlUtils: HtmlUtils,
+                restrictEditing: this.restrictEditing || false
```
Do we expect that sometimes restrictEditing won't be provided?

staubina replied:
probably an overprotection

cahrens commented on cms/static/js/views/content_group_list.js:
```diff
@@ -19,6 +19,10 @@ define([
 
+        restrictEditing: false,
+
+        scheme: 'none',
```
Why is scheme None?

staubina replied:
Meant to be an init value, there are better options.

cahrens commented on cms/static/js/views/list.js:
```diff
@@ -11,6 +11,8 @@
  *   Note that it must be translated.
  * - emptyMessage (string): Text to render when the list is empty.
+ * - restrictEditing (bool) : Boolean flag for whether or not to restrict and hide editing options.
+ * - scheme (string): Scheme name for content group scheme
```
js/views/list is a generic list view. It shouldn't have the notion of a "scheme".

staubina replied:
This was a missed remnant of a previous change, I will look at removing it.

cahrens commented on cms/static/js/views/list_item.js:
```diff
@@ -22,6 +22,7 @@ define([
         canDelete: false,
 
         initialize: function() {
+            this.restrictEditing = this.options.restrictEditing || false;
```
Can you write up a description of how/when default values are provided for restrictEditing? For example, here a default value of false is supplied, but in list.js there is not a default value (and the documentation doesn't say that it is optional).

staubina replied:
The default values are here in case this object is used elsewhere.  I tried to make it so that restrictEditing was not a required option.  This resulted in me not having to change all usages of these classes.

cahrens commented on cms/static/js/views/pages/group_configurations.js:
```diff
@@ -6,6 +6,11 @@ function($, _, gettext, BasePage, GroupConfigurationsListView, ContentGroupListV
     'use strict';
     var GroupConfigurationsPage = BasePage.extend({
         initialize: function(options) {
+            var restrictEditing = false,
```
I would get rid of the default value here for restrictEditing.

cahrens commented on cms/static/js/views/pages/group_configurations.js:
```diff
@@ -14,18 +19,38 @@ function($, _, gettext, BasePage, GroupConfigurationsListView, ContentGroupListV
+                this.allGroupViewList.push(
+                    new ContentGroupListView({
+                        collection: this.allGroupConfigurations[i].get('groups'),
+                        restrictEditing: restrictEditing,
```
You can simplify this by ditching the variable restrictEditing all together, and doing this:

````
restrictEditing: this.allGroupConfigurations[i].get('scheme') === this.ENROLLMENT_SCHEME

cahrens commented on cms/static/js/views/pages/group_configurations.js:
```diff
@@ -38,8 +63,22 @@ function($, _, gettext, BasePage, GroupConfigurationsListView, ContentGroupListV
+         * Checks the Content Group Configurations to see if the isDirty bit is set
+         * @returns {boolean} True if any content group has the dirty bit set.
+         */
+        areContentConfigurationsDirty: function() {
```
Here you are iterating through all GroupConfigurations, correct?

staubina replied:
Until I hit the first instance of isDirty being true.

staubina replied:
Ah I see, I will rename.

cahrens commented on cms/templates/js/content-group-details.underscore:
```diff
@@ -23,17 +23,19 @@
    </ol>
 
```
Rename this file to... something. :)

cahrens commented on common/test/acceptance/tests/test_cohorted_courseware.py:
```diff
@@ -33,6 +60,11 @@ def setUp(self, is_staff=True):
 
         self.content_group_a = "Content Group A" + XSS_INJECTION
         self.content_group_b = "Content Group B" + XSS_INJECTION
+        self.content_group_verified = "Verified Track"
```
verified_enrollment_track, or verified_enrollment_group, or verified_group?

cahrens commented on common/test/acceptance/tests/test_cohorted_courseware.py:
```diff
@@ -115,20 +173,26 @@ def create_content_groups(self):
             problem.edit_visibility()
             visibility_dialog = ComponentVisibilityEditorView(self.browser, problem.locator)
-            visibility_dialog.select_groups_in_partition_scheme(visibility_dialog.CONTENT_GROUP_PARTITION, groups)
+            partition_name = visibility_dialog.CONTENT_GROUP_PARTITION
```
Personally, I find this easier to read, though perhaps it is too long. If so, I would go with an if/else.

````
partition_name = visibility_dialog.ENROLLMENT_TRACK_PARTITION if group_partition == enrollment_group else visibility_dialog.CONTENT_GROUP_PARTITION

cahrens commented on common/test/acceptance/tests/test_cohorted_courseware.py:
```diff
@@ -150,20 +214,39 @@ def add_cohort_with_student(cohort_name, content_group, student):
             verify_expected_problem_visibility(self, courseware_page, expected_problems)
 
+        def set_user_track(track=None):
+            payment_and_verification_flow = PaymentAndVerificationFlow(self.browser, self.course_id)
```
If this same code exists someplace else, please create a helper method that multiple tests can reference.

cahrens commented on openedx/core/djangoapps/verified_track_content/partition_scheme.py:
```diff
@@ -45,14 +45,6 @@ def groups(self):
-
     def from_json(self):
         """
         Because this partition is dynamic, to_json and from_json are not supported.
```
Update comment-- to_json is now supported, but shouldn't be used to persist this partition within the course itself (used by Studio for sending data to front-end code).

cahrens commented on common/test/acceptance/tests/test_cohorted_courseware.py:
```diff
@@ -177,31 +260,87 @@ def login_and_verify_visible_problems(username, email, expected_problems):
+
+class EnrollmentTrackModeTest(ContainerBase):
+
+    def setUp(self, is_staff=True):
```
Can you add just a brief comment to this test (don't need to do the BDD style) stating that the purpose of these tests are to ensure that Enrollment Tracks only show on the Group Configurations page if there are > 1 tracks?

cahrens commented:
I'm done with this pass. I do think we should rename files like "content_group_details" and "content_group_list". I vote for "partition_group_details" or "group_configuration_group_details" or "configuration_group_details".

jlajoie commented on common/test/acceptance/tests/test_cohorted_courseware.py:
```diff
@@ -33,6 +60,11 @@ def setUp(self, is_staff=True):
         self.content_group_a = "Content Group A" + XSS_INJECTION
         self.content_group_b = "Content Group B" + XSS_INJECTION
+        self.content_group_verified = "Verified Track"
+        self.content_group_audit = "Audit Track"
```
Is there a reason these aren't using the constants?

jlajoie commented on common/test/acceptance/tests/test_cohorted_courseware.py:
```diff
@@ -150,20 +214,39 @@ def add_cohort_with_student(cohort_name, content_group, student):
+             self.group_audit_problem,
+             self.group_a_and_b_problem,
+             self.visible_to_all_problem
+             ],
```
Nit: trailing comma

catong commented on cms/templates/group_configurations.html:
```diff
@@ -47,32 +49,45 @@ <h1 class="page-header">
+      % if should_show_enrollment_track:
+        <div class="bit">
+          <div class="enrollment-track-doc">
+            <h3 class="title-3">${_("Enrollment Track Content Groups")}</h3>
```
Enrollment Track Groups

Enrollment track groups are available only if you have more than one enrollment track in your course. Learners enrolled in each enrollment track in your course are automatically included in the corresponding enrollment track group. 

Enrollment track groups allow you to offer different course content to the learners in each enrollment track. In the course outline, you can designate some content as visible only to learners in a specific enrollment track. 

You cannot edit enrollment track groups, but you can expand each group to view details of the course content that is designated for learners in the group.

catong replied:
@sstack22 @staubina @cahrens Thoughts about the terminology that I'm suggesting? ^ I initially liked referring to "course mode" but I'm not sure if "course mode" is surfaced to course teams. Sticking with the single concept of "enrollment track" might keep things simpler, but does sound kind of repetitive.

staubina replied:
@catong I made the requested changes, Sandbox updating now.

cahrens replied:
@catong I don't think we need to say "Enrollment track groups are available only if you have more than one enrollment track in your course". This whole sidebar documentation will only show (at least as implemented) if there are multiple enrollment tracks in the course. Did you intend it to always show?

cahrens replied:
@catong We can't yet edit on the course outline (at the unit level)-- we can only do it at the component level on the unit page. I'm reacting to this--

" In the course outline, you can designate some content as visible only to learners in a specific enrollment track."

catong replied:
@cahrens Yes, I knew the section would only show up if more than one enrollment track existed. So I agree, we don't have to include that info in the sidebar text.
@staubina Let's remove the first two sentences, so that the text starts with "Enrollment track groups allow you to offer..."
Also, can we change the heading above the enrollment track boxes to read "Enrollment Track Groups" instead of "Enrollment Tracks"?

catong replied:
@staubina Revised:
Enrollment Track Groups

Enrollment track groups allow you to offer different course content to learners in each enrollment track. Learners enrolled in each enrollment track in your course are automatically included in the corresponding enrollment track group.

On unit pages in the course outline, you can designate components as visible only to learners in a specific enrollment track.

You cannot edit enrollment track groups, but you can expand each group to view details of the course content that is designated for learners in the group.

catong commented on cms/templates/group_configurations.html:
```diff
@@ -47,32 +49,45 @@ <h1 class="page-header">
+            <p><a href="${get_online_help_info(enrollment_track_help_token())['doc_url']}" target="_blank" class="button external-help-button">${_("Learn More")}</a></p>
+          </div>
+        </div>
+      % endif
```
@staubina I'm not able to add comments at the exact lines, but in the sidebar text for Content Groups and Experiment Groups, can we change "students" to "learners"? I think there are just 2 instances: line 95 and line 104. Thanks!

staubina commented:
jenkins run js

staubina commented on cms/envs/common.py:
```diff
@@ -229,7 +229,7 @@
 
     # Whether or not the dynamic EnrollmentTrackUserPartition should be registered.
-    'ENABLE_ENROLLMENT_TRACK_USER_PARTITION': False,
+    'ENABLE_ENROLLMENT_TRACK_USER_PARTITION': True,
```
TODO: Remove this before Merging.

staubina commented on cms/templates/js/configuration-group-details.underscore:
```diff
@@ -0,0 +1,74 @@
+<div class="collection-details wrapper-group-configuration">
```
Note: I renamed this file.  I will try to highlight changes.

cahrens replied:
Rename as partition-group-details.

staubina commented on cms/templates/js/configuration-group-details.underscore:
```diff
@@ -0,0 +1,74 @@
+  <div class="collection-references wrapper-group-configuration-usages">
+    <% if (!_.isEmpty(usage)) { %>
+      <h4 class="intro group-configuration-usage-text">
+        <%= HtmlUtils.interpolateHtml(gettext('This {groupType} is used in:'), {groupType: groupTypeName}) %>
```
Note: Added {groupType} so that this string would be dynamic

cahrens replied:
I would prefer to avoid this for simplicity. How about just "Used in"?

cahrens replied:
If you are going to do this, you might as well just check the scheme name (as you did below on line 58) instead of passing in groupTypeName. 

However, I would prefer this file not have any scheme-specific code (if we can get rid of line 62).

staubina commented on cms/templates/js/configuration-group-details.underscore:
```diff
@@ -0,0 +1,74 @@
+    <% } else { %>
+        <p class="group-configuration-usage-text">
+            <% if (scheme == 'enrollment_track') { %>
+                <%= HtmlUtils.interpolateHtml(gettext('This enrollment track is not in use.'))%>
```
Note: Added a conditional so that the string would reflect the partition scheme.

catong replied:
If we are calling these things "enrollment track groups" can we also update the string here?

cahrens replied:
How about just "Not in use"?

staubina commented on common/test/acceptance/pages/common/utils.py:
```diff
@@ -100,3 +102,22 @@ def hover(browser, element):
     """
     ActionChains(browser).move_to_element(element).perform()
+
+
```
Note: New utility method to support user track enrollment.

staubina commented on lms/envs/common.py:
```diff
@@ -373,7 +373,7 @@
     'ENABLE_COOKIE_CONSENT': False,
 
     # Whether or not the dynamic EnrollmentTrackUserPartition should be registered.
-    'ENABLE_ENROLLMENT_TRACK_USER_PARTITION': False,
```
TODO: Remove before merging!

staubina commented on lms/static/sass/course/wiki/_wiki.scss:
```diff
@@ -135,6 +135,7 @@
   .entry-title {
     padding-bottom: 8px;
     margin-bottom: 22px;
+    margin-top: 0;
```
Minor out of scope fix for the Wiki tabs header.

staubina commented:
@cahrens @catong @jlajoie 

Please take a look.  I believe I have addressed all comments.  There were quite a few so please let me know if something was missed.  

The sandbox should be updated soon.

cahrens commented:
You forgot to remove "Content" from the title.

![image](https://cloud.githubusercontent.com/assets/484484/25016046/75187534-204c-11e7-8975-7b8884f8ff9b.png)

staubina commented:
Sandbox is updated.

cahrens commented on cms/djangoapps/contentstore/course_group_config.py:
```diff
@@ -200,8 +201,6 @@ def _get_content_groups_usage_info(course, items):
         """
         Returns all units names and their urls.
```
I would rename this method.

cahrens commented on cms/djangoapps/contentstore/course_group_config.py:
```diff
@@ -363,3 +363,21 @@ def get_or_create_content_group(store, course):
+    @staticmethod
+    def get_all_user_partitions(store, course):
+        """
+        Returns all the available content groups with updated group usage information
```
Update doc.

And the name of the method is confusing with "get_all_partitions_for_course", etc. What is special about this method? Is it that it returns JSON and has the usage information?

How about get_all_user_partition_details or get_all_partitions_with_usage?

cahrens commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -1521,15 +1521,32 @@ def group_configurations_list_handler(request, course_key_string):
+                        all_partitions.insert(0, partition)
+
+            # Add empty content group if list returned is empty.  This will add ability to add new groups in the view.
+            if len(all_partitions) == 0 or COHORT_SCHEME not in group_schemes:
```
Change to just COHORT_SCHEME not in group_schemes.

cahrens commented on cms/djangoapps/contentstore/views/tests/test_group_configurations.py:
```diff
@@ -613,6 +613,16 @@ class GroupConfigurationsUsageInfoTestCase(CourseTestCase, HelperMethods):
 
+    def _get_user_partition(self, scheme):
+        """
+        Returns the content group configuration with the specified scheme.
```
Update doc.

staubina commented:
jenkins run js

cahrens commented on cms/static/js/spec/views/group_configuration_spec.js:
```diff
@@ -775,13 +775,13 @@ define([
         beforeEach(function() {
             TemplateHelpers.installTemplates(
-                ['content-group-editor', 'content-group-details', 'list']
+                ['content-group-editor', 'configuration-group-details', 'list']
```
Nit: I would prefer you standardize on either "partition-group" or "configuration-group" (see line 778 above).

staubina replied:
We discussed this in person.  Please let me know if you would still like more changes to this.

cahrens commented on cms/static/js/views/list.js:
```diff
@@ -11,6 +11,7 @@
  *   of items this list contains.  For example, 'Group Configuration'.
  *   Note that it must be translated.
  * - emptyMessage (string): Text to render when the list is empty.
+ * - restrictEditing (bool) : Boolean flag for whether or not to restrict and hide editing options.
```
Please add that it default to false if not provided.

cahrens commented on cms/static/js/views/partition_group_details.js:
```diff
@@ -21,29 +21,39 @@ define([
                 'collection',
-                'content-group-details',
-                'content-group-details-' + index
+                'configuration-group-details',
```
Probably change this to partition.

cahrens commented on cms/static/js/views/partition_group_details.js:
```diff
@@ -21,29 +21,39 @@ define([
-            this.template = this.loadTemplate('content-group-details');
+            this.template = this.loadTemplate('configuration-group-details');
+            this.restrictEditing = this.options.restrictEditing || false;
+            this.scheme = this.options.scheme || 'cohort';
```
I don't think it makes sense to have a default scheme (I also question having a default value for restrictEditing).

cahrens commented on cms/static/js/views/partition_group_details.js:
```diff
@@ -21,29 +21,39 @@ define([
 
         render: function(showContentGroupUsages) {
-            var attrs = $.extend({}, this.model.attributes, {
+            var groupTypeName = 'content group',
```
Can you just use the scheme? I'd like this code not to know about Content Groups vs. Enrollment Tracks.

cahrens replied:
I just looked at how this is used, and I'd like to avoid introducing this complexity (also note groupTypeName is not translated).

cahrens commented on cms/templates/js/configuration-group-details.underscore:
```diff
@@ -0,0 +1,74 @@
+                <%= HtmlUtils.interpolateHtml(gettext('This enrollment track is not in use.'))%>
+            <%} else {%>
+                <%= HtmlUtils.interpolateHtml(
+                    gettext('This content group is not in use. Add a content group to any unit from the {linkStart}Course Outline{linkEnd}.'),
```
@catong Is it possible to make this message generic? Really, it applies to Enrollment Tracks just as much.

cahrens commented on cms/static/js/views/partition_group_list.js:
```diff
@@ -0,0 +1,33 @@
+        className: 'partition-group-list',
+
+        // Translators: This refers to a content group that can be linked to a student cohort.
+        itemCategoryDisplayName: gettext('content group'),
```
Uck, these "content group" messages aren't very nice in this class. However, I suppose this can be cleaned up if we ever have another editable type of UserPartition/GroupConfiguration.

cahrens commented on cms/templates/group_configurations.html:
```diff
@@ -16,7 +17,7 @@
 
 <%block name="header_extras">
-% for template_name in ["group-configuration-details", "group-configuration-editor", "group-edit", "content-group-editor", "content-group-details", "basic-modal", "modal-button", "list"]:
+% for template_name in ["group-configuration-details", "group-configuration-editor", "group-edit", "content-group-editor", "configuration-group-details", "basic-modal", "modal-button", "list"]:
```
partition-group-details

cahrens commented on common/test/acceptance/tests/lms/test_lms.py:
```diff
@@ -444,17 +443,7 @@ def test_immediate_verification_enrollment(self):
         # Create a user and log them in
         student_id = AutoAuthPage(self.browser).visit().get_user_id()
 
-        # Navigate to the track selection page
```
Ahhhh, doesn't this make you feel good?

cahrens commented:
jenkins run python

cahrens commented on cms/djangoapps/contentstore/course_group_config.py:
```diff
@@ -187,21 +188,19 @@ def _get_content_experiment_usage_info(store, course, split_tests):  # pylint: d
 
     @staticmethod
-    def _get_content_groups_usage_info(course, items):
+    def _get_partition_usage_info(course, items):
```
Oh dear-- we now have a method called "get_partitions_usage_info" that calls "_get_partition_usage_info".

Since _get_partition_usage_info is only used in one place, can you just move its implementation into get_partitions_usage_info and delete the helper method?

staubina replied:
This was pre existing.  And yes I can merge them.

cahrens commented on cms/djangoapps/contentstore/course_group_config.py:
```diff
@@ -187,21 +188,19 @@ def _get_content_experiment_usage_info(store, course, split_tests):  # pylint: d
-    def get_content_groups_usage_info(store, course):
+    def get_partitions_usage_info(store, course):
         """
         Get usage information for content groups.
```
Update doc string.

cahrens commented on cms/djangoapps/contentstore/course_group_config.py:
```diff
@@ -282,22 +281,23 @@ def _get_content_groups_items_usage_info(course, items):
-    def _iterate_items_and_content_group_ids(course, items):
+    def _iterate_items_and_group_ids(course, items):
         """
         Iterate through items and content group IDs in a course.
```
Update doc string.

cahrens commented on cms/djangoapps/contentstore/course_group_config.py:
```diff
@@ -363,3 +363,21 @@ def get_or_create_content_group(store, course):
+    @staticmethod
+    def get_all_user_partition_details(store, course):
+        """
+        Returns all the available content groups with updated group usage information
```
Update doc string.

cahrens commented on cms/djangoapps/contentstore/course_group_config.py:
```diff
@@ -363,3 +363,21 @@ def get_or_create_content_group(store, course):
+        all_partitions = get_all_partitions_for_course(course)
+        all_configurations = []
+        for partition in all_partitions:
+            configuration = GroupConfiguration.update_content_group_usage_info(
```
Looks like update_content_group_usage_info did not get renamed?

cahrens commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -1521,15 +1521,32 @@ def group_configurations_list_handler(request, course_key_string):
+                    if should_show_enrollment_track:
+                        all_partitions.insert(0, partition)
+
+            # Add empty content group if list returned is empty.  This will add ability to add new groups in the view.
```
Update doc string about when the empty content group is added.

cahrens commented on cms/templates/js/partition-group-details.underscore:
```diff
@@ -23,25 +23,29 @@
     <% if (!_.isEmpty(usage)) { %>
-      <h4 class="intro group-configuration-usage-text"><%- gettext('This content group is used in:') %></h4>
+      <h4 class="intro group-configuration-usage-text">
+        <%= gettext('This group controls visibility of:') %>
```
Should be <%- instead of <%=. The latter doesn't do escaping, and should only be used with HTML (which has already been run through HtmlUtils).

cahrens commented on cms/templates/js/partition-group-details.underscore:
```diff
@@ -52,15 +56,18 @@
     <% } else { %>
         <p class="group-configuration-usage-text">
             <%= HtmlUtils.interpolateHtml(
```
<%= is correct in this case.

cahrens commented on openedx/core/djangoapps/verified_track_content/tests/test_partition_scheme.py:
```diff
@@ -53,10 +53,9 @@ def test_multiple_groups(self):
         user_partition = create_enrollment_track_partition(self.course)
-        with self.assertRaises(TypeError):
-            user_partition.to_json()
+        self.assertIsNotNone(user_partition.to_json())
```
Instead of just asserting that it isn't None, can you test something about what it should be?

I would call it test_to_json, and assert what you expect the output to be.

cahrens commented on openedx/core/djangoapps/verified_track_content/partition_scheme.py:
```diff
@@ -45,17 +45,11 @@ def groups(self):
         """
-        Because this partition is dynamic, to_json and from_json are not supported.
+        Because this partition is dynamic, from_json is not supported.
+        to_json is now supported, but shouldn't be used to persist this partition
```
Nit: remove "now".

cahrens commented on common/test/acceptance/tests/test_cohorted_courseware.py:
```diff
@@ -177,31 +252,98 @@ def login_and_verify_visible_problems(username, email, expected_problems):
+
+        login_and_verify_visible_problems(
+            self.cohort_audit_student_username, self.cohort_audit_student_email,
+            [self.group_audit_problem, self.visible_to_all_problem],
```
I'm confused why cohort_a_student_username and cohort_b_student_username don't see the group_audit_problem (lines 246 and 251), as they should be in the default course mode.

staubina replied:
They default to a track of honor.

cahrens replied:
Odd, I would have expected the default mode to be Audit. But if it is Honor, can you please change the course to have Honor and Verified as the 2 course modes (or enroll these other students in Audit)? For someone just reading the tests, it's very confusing to see that some students see neither the Audit nor the Verified content (since this is not a case that happens in "real life").

cahrens commented on common/test/acceptance/tests/test_cohorted_courseware.py:
```diff
@@ -177,31 +252,98 @@ def login_and_verify_visible_problems(username, email, expected_problems):
         self.view_cohorted_content_as_different_users()
+
+
+class EnrollmentTrackModeTest(ContainerBase):
```
I think I would move this test to test_studio_settings.py as a peer of ContentGroupConfigurationTest. Also add a shard declaration.

cahrens commented:
@staubina I'm done with this pass. Bok choy tests indicate that something isn't quite right.

staubina commented:
jenkins run a11y

staubina commented:
jenkins run bokchoy

staubina commented:
@cahrens @jlajoie @catong 

I am not sure what is going on with the test runs.  I am checking with test eng.  I have updated the Sandbox with last nights changes though.

staubina commented:
jenkins run bokchoy

cahrens commented on common/lib/xmodule/xmodule/partitions/partitions_service.py:
```diff
@@ -74,7 +74,7 @@ def _create_enrollment_track_partition(course):
     partition = enrollment_track_scheme.create_user_partition(
         id=ENROLLMENT_TRACK_PARTITION_ID,
-        name=_(u"Enrollment Tracks"),
+        name=_(u"Enrollment Track Groups"),
```
@catong Note that this will impact what is shown in the visibility dialog as well.

cahrens commented on common/test/acceptance/pages/common/utils.py:
```diff
@@ -100,3 +103,44 @@ def hover(browser, element):
+    """
+    audit_track = "Audit Track"
+    verified_track = "Verified Track"
+    for track in tracks:
```
Hmmm, if this is going to be a general utility method, I don't like how "tracks" can only contain "Audit Track" or "Verified Track" (and there is no documentation that indicates what "tracks" can be. I felt squeamish about it before, but it was in a single test file.

I think that "tracks" should allow any of the mode slugs ("audit", "verified", "honor", etc.).

cahrens commented on common/test/acceptance/pages/lms/track_selection.py:
```diff
@@ -34,24 +34,24 @@ def is_browser_on_page(self):
 
             Keyword Arguments:
-                mode (str): Can be "honor" or "verified"
+                mode (str): Can be "audit" or "verified"
```
Why did you change from "honor" to "audit"?

staubina replied:
"audit" is not available in the track selection, "audit" is.  Is this incorrect?

cahrens replied:
Oh, you mean the actual track selection page? Yes, "honor" is a legacy mode and likely not available. So it's fine to switch it to "audit", but let's enroll those other users in "audit".

So this means that nobody was calling this method with "honor", as it didn't work?

cahrens commented on common/test/acceptance/tests/test_cohorted_courseware.py:
```diff
@@ -34,6 +37,9 @@ def setUp(self, is_staff=True):
         self.content_group_b = "Content Group B" + XSS_INJECTION
 
+        # Creates the Course modes needed to test enrollment tracks
+        add_enrollment_course_modes(self, [AUDIT_TRACK, VERIFIED_TRACK])
```
As I mentioned on the helper method, I think these should just be ["audit", "verified"] (slugs).

cahrens commented on common/test/acceptance/tests/test_cohorted_courseware.py:
```diff
@@ -48,6 +54,26 @@ def setUp(self, is_staff=True):
+        ).visit()
+
+        # Create audit student
+        self.cohort_audit_student_username = "cohort_audit_student"
```
As mentioned before, let's make sure all learners are in a track so that the test more closely mimics reality.

cahrens replied:
@staubina Thinking about this overnight (OK, I promise I didn't give it TOO much thought overnight), another alternative that I think is acceptable is to add comments in the methods that check which components are visible to the "cohort" students. Have the comments state that the learner is auto-enrolled in the Honor track, and therefore can't see the Audit (or Verified) content.

Although it doesn't mimic "reality", at least it helps explain the expected/tested behavior for the next person reading the code.

cahrens commented on cms/static/js/views/pages/group_configurations.js:
```diff
@@ -1,11 +1,16 @@
         initialize: function(options) {
+            var currentScheme,
+                i;
+            this.COHORT_SCHEME = 'cohort';
```
COHORT_SCHEME is unused-- I will delete it.

cahrens commented on cms/static/js/views/pages/group_configurations.js:
```diff
@@ -1,11 +1,16 @@
+            var currentScheme,
+                i;
+            this.COHORT_SCHEME = 'cohort';
+            this.ENROLLMENT_SCHEME = 'enrollment_track';
```
There is no reason that this should be a class variable-- I will make it a local variable.

staubina commented:
:+1: for the current set of changes made by @cahrens

cahrens commented:
👍 from me. @jlajoie and @catong, can you take a last look at this PR? I'm hoping to merge it tomorrow (I will remove the temporary commit first and squash).

cahrens commented:
jenkins run js

catong commented:
@cahrens 👍 I can update the Help mapping in a separate future PR after the new documentation is published and the link can resolve correctly.

jlajoie commented:
:+1: lgtm
